build-logic/basics/src/main/kotlin/gradlebuild/basics/BuildEnvironment.kt

```diff
@@ -97,6 +109,11 @@ object BuildEnvironment {
     val isTravis = "TRAVIS" in System.getenv()
     val isJenkins = "JENKINS_HOME" in System.getenv()
     val isGhActions = "GITHUB_ACTIONS" in System.getenv()
+    val isCodeQl: Boolean by lazy {
+        // This logic is kept here instead of `codeql-analysis.init.gradle` because that file will hopefully be removed in the future.
+        // Removing that file is waiting on the GitHub team fixing an issue in Autobuilder logic.
+        CODEQL_ENVIRONMENT_VARIABLES.any { it in System.getenv() }
```

Comment:
```suggestion
        CODEQL_ENVIRONMENT_VARIABLES.any { System.getenv(it) != null }
```

I am pretty sure my suggestion doesn't cause the lookup of all env vars. I don't know if your version also does this, maybe @mlopatkin might have an idea.

Reply:
Both should be equivalent as Map's `in` seems to be delegating to `containsKey` in Kotlin.

Reply:
The rest of the file is using the `"var" in System.getenv()` pattern.

https://github.com/gradle/gradle/blob/3f2bdeaff4720742dfedf9fd0877a9754a95b720/build-logic/basics/src/main/kotlin/gradlebuild/basics/BuildEnvironment.kt#L95-L99

Also, regardless of if it does a lookup of all the system environment variables or not, is there actually any coupling here that's happening that could actually invalidate any kind of cache?

>  If you check all environment variables here, I suppose all of them will be an input to the build logic

Could you clarify what you mean by this? I wasn't aware that Gradle did any intercepting the lookup of system environment variables and tracking that as "input" to the Gradle build. If so, do we document this somewhere? If Gradle isn't implicitly tracking the lookup of environment variables as tracked "inputs", isn't this entire exercise of changing from what I originally proposed an exercise in premature optimization? Both my original proposal (`System.getEnv().keys.any { it.startsWith("CODEQL_") }`) and the change to only look at specific environment variables both operate in `O(1)` because the set of keys that are held in the environment variable Map are of constant size.

Regardless, this is how `containsKey` is implemented in `HashMap`:
```java
public boolean containsKey(Object key) {
        return getNode(hash(key), key) != null;
    }
```
As such, `System.getenv(it) != null` would be exactly the same as `it in System.getenv()`.

Reply:
Awesome, so `it in System.getenv()` doesn't cause us to detect the querying of all environment variables. So I suppose your solution is good already.

Gradle intercepts access to environment variables during configuration time since this PR: https://github.com/gradle/gradle/pull/18784. There is a little bit of documentation here, though not in particular on this auto-detection: https://docs.gradle.org/nightly/userguide/configuration_cache.html#config_cache:requirements:undeclared_env_var_read

If you use `System.getEnv().keys.any { it.startsWith("CODEQL_") }`, then the build script will go through all environment variables in `getEnv()` until if finds one starting with `CODEQL_`. All the checked environment variables will be part of the configuration cache key. If there aren't any `CODEQL_` env vars (e.g. for developer builds), then all environment variables will be a build input, since all names needed to be checked. This is what the change to check only a given list avoids: Only those env vars (or their existence) are an input to the build, not all env vars.

Reply:
Perfect! Thanks for the clarification!